File: shoring/blackjackEngine/src/main/java/com/panopset/blackjackEngine/CycleSnapshot.java
package com.panopset.blackjackEngine;

import java.util.ArrayList;
import java.util.List;

public class CycleSnapshot {

  private final Bankroll bankroll;
  private final HandDealer dealer;
  private final List<Player> players = new ArrayList<>();
  private final Metrics metrics;
  private final int nextBet;
  private final String mistakeHeader;
  private final String mistakeMessage;
  private final String dealerMessage;
  private final List<String> gameStatusVertical = new ArrayList<>();
  private final String gameStatusHorizontal;
  private final List<String> statusChipsVertical = new ArrayList<>();
  private final String statusChipsHorizontal;
  private boolean isPainted = false;
  private final String action;

  CycleSnapshot(final BlackjackGameEngine bge, String action) {
    this.action = action;
    bankroll = new Bankroll(bge.getBankroll());
    dealer = new HandDealer(bge.getCycle().getDealer());
    for (Player player : bge.getCycle().getPlayers()) {
      players.add(new Player(player));
    }
    metrics = new Metrics(bge.getMetrics());
    this.nextBet = bge.getNextBet();
    this.mistakeHeader = bge.getMistakeHeader();
    this.mistakeMessage = bge.getMistakeMessage();
    this.dealerMessage = bge.getDealerMessage();
    for (String gsv : bge.getGameStatusVertical()) {
      gameStatusVertical.add(gsv);
    }
    this.gameStatusHorizontal = bge.getGameStatusHorizontal();
    for (String scv : bge.getStatusChipsVertical()) {
      this.statusChipsVertical.add(scv);
    }
    this.statusChipsHorizontal = bge.getStatusChipsHorizontal();
  }

  public Bankroll getBankroll() {
    return bankroll;
  }

  public HandDealer getDealer() {
    return dealer;
  }

  public List<Player> getPlayers() {
    return players;
  }

  public Metrics getMetrics() {
    return metrics;
  }

  public int getNextBet() {
    return nextBet;
  }

  public String getMistakeHeader() {
    return mistakeHeader;
  }

  public String getMistakeMessage() {
    return mistakeMessage;
  }

  public String getDealerMessage() {
    return dealerMessage;
  }

  public List<String> getGameStatusVertical() {
    return gameStatusVertical;
  }

  public String getGameStatusHorizontal() {
    return gameStatusHorizontal;
  }

  public List<String> getStatusChipsVertical() {
    return statusChipsVertical;
  }

  public String getStatusChipsHorizontal() {
    return statusChipsHorizontal;
  }

  public void setPainted() {
    isPainted = true;
  }
  
  public boolean isPainted() {
    return isPainted;
  }
  
  public String getAction() {
    return action;
  }
}

File: code/shoring/blackjackEngine/src/main/kotlin/com/panopset/blackjackEngine/CycleSnapshot.kt
package com.panopset.blackjackEngine

class CycleSnapshot(private val bgs: BlackjackGameState) {

    var isPainted = false
        private set

    fun getPlayers(): List<Player> {
        return bgs.players
    }

    fun getGameStatusVertical(): List<String> {
        return bgs.gameStatusVertical
    }

    fun getStatusChipsVertical(): List<String> {
        return bgs.statusChipsVertical
    }

    fun setPainted() {
        isPainted = true
    }

    fun getAction(): String {
        return bgs.action
    }

    fun getStakeIncludingHands(): Long {
        return bgs.stakeIncludingHands
    }

    fun getChips(): Long {
        return bgs.chips
    }

    fun getMetrics(): Metrics {
        return bgs.metrics
    }

    fun getNextBet(): Int {
        return bgs.nextBet
    }

    fun getDealer(): HandDealer {
        return bgs.handDealer
    }

    fun getMistakeHeader(): String {
        return bgs.mistakeHeader
    }

    fun getMistakeMessage(): String {
        return bgs.mistakeMessage
    }

    fun getDealerMessage(): String {
        return bgs.dealerMessage
    }

    fun getGameStatusHorizontal(): String{
        return bgs.gameStatusHorizontal
    }

    fun getStatusChipsHorizontal(): String {
        return bgs.statusChipsHorizontal
    }
}
